pkg/trisa/envelope/envelope.go

```diff
@@ -0,0 +1,751 @@
+
+// ID returns the envelope ID
+func (e *Envelope) ID() string {
+	return e.msg.Id
+}
+
+// Proto returns the trisa.SecureEnvelope protocol buffer.
+func (e *Envelope) Proto() *api.SecureEnvelope {
+	return e.msg
+}
+
+// Payload returns the parsed trisa.Payload protocol buffer if available. If the
+// envelope is not decrypted then an error is returned.
+func (e *Envelope) Payload() (_ *api.Payload, err error) {
+	state := e.State()
+	if state != Clear && state != ClearError {
+		err = fmt.Errorf("envelope is in state %q: payload may be invalid", state)
+	}
+	return e.payload, err
+}
+
+// Error returns the rejection error on the envelope if it exists
+func (e *Envelope) Error() *api.Error {
+	// Ensure a nil error is returned if the error is zero-valued
+	if e.msg.Error != nil && e.msg.Error.IsZero() {
+		return nil
+	}
+	return e.msg.Error
+}
+
+// Timestamp returns the ordering timestamp of the secure envelope. If the timestamp is
+// not on the envelope or it cannot be parsed, an error is returned.
+func (e *Envelope) Timestamp() (ts time.Time, err error) {
+	if e.msg.Timestamp == "" {
+		return ts, &api.Error{Code: api.BadRequest, Message: "missing ordering timestamp on secure envelope"}
+	}
+
+	// First attempt: parse with nanosecond resolution
+	if ts, err = time.Parse(time.RFC3339Nano, e.msg.Timestamp); err != nil {
+		// Second attempt: try without nanosecond resolution
+		if ts, err = time.Parse(time.RFC3339, e.msg.Timestamp); err != nil {
+			return ts, &api.Error{Code: api.BadRequest, Message: "could not parse ordering timestamp on secure envelope as RFC3339 timestamp"}
+		}
+	}
+	return ts, err
+}
+
+// Crypto returns the cryptographic method used to encrypt/decrypt the payload.
+func (e *Envelope) Crypto() crypto.Crypto {
+	return e.crypto
+}
+
+// Sealer returns the cryptographic cipher method used to seal/unseal the envelope.
+func (e *Envelope) Sealer() crypto.Cipher {
+	return e.seal
+}
+
+//===========================================================================
+// Envelope Validation
+//===========================================================================
+
+// State returns the current state of the envelope.
+func (e *Envelope) State() State {
+	// If a payload exists on the envelope, then it is in the clear
+	if e.payload != nil {
+		if e.msg.Error == nil || e.msg.Error.IsZero() {
+			return Clear
+		}
+		return ClearError
+	}
+
+	// If there is no payload, there should be a secure envelope
+	if e.msg == nil {
+		return Unknown
+	}
+
+	// If there is a secure envelope, it should be valid
+	if err := e.ValidateMessage(); err != nil {
+		return Corrupted
+	}
+
+	// Check if the envelope is marked as sealed
+	if e.msg.Sealed {
+		if e.msg.Error == nil || e.msg.Error.IsZero() {
+			return Sealed
+		}
+		return SealedError
+	}
+
+	// Message is unsealed
+	if e.msg.Error == nil || e.msg.Error.IsZero() {
+		return Unsealed
+	}
+	return UnsealedError
```

Comment:
Yes, if it had unsealed payload data and an error -- this is not something that this particular library will produce, but it is a state that should be handled by the user in case they receive an envelope that was echoed back with an error but without zeroing the payload.